src/writer.rs

```diff
@@ -804,14 +845,57 @@ mod tests {
     fn test_overflow_subtract() {
         let overflow_size = std::u32::MAX / size_of::<FdtReserveEntry>() as u32 - 3;
         let too_large_mem_reserve: Vec<FdtReserveEntry> = (0..overflow_size)
-            .map(|i| FdtReserveEntry {
-                address: u64::from(i) * 2,
-                size: 1,
-            })
+            .map(|i| FdtReserveEntry::new(u64::from(i) * 2, 1).unwrap())
             .collect();
         let mut fdt = FdtWriter::new(&too_large_mem_reserve).unwrap();
         let root_node = fdt.begin_node("").unwrap();
         fdt.end_node(root_node).unwrap();
         assert_eq!(fdt.finish().unwrap_err(), Error::TotalSizeTooLarge);
     }
+
+    #[test]
+    fn test_invalid_mem_reservations() {
+        // Test that we cannot create an invalid FDT reserve entry where the
+        // end address of the region would not fit in an u64.
+        assert_eq!(
+            FdtReserveEntry::new(0x1, u64::MAX).unwrap_err(),
+            Error::InvalidMemoryReservation
+        )
+    }
+
+    #[test]
+    fn test_cmp_mem_reservations() {
+        // Test that just the address is taken into consideration when comparing to `FdtReserveEntry`.
+        assert_eq!(
+            FdtReserveEntry::new(0x1, 10)
+                .unwrap()
+                .cmp(&FdtReserveEntry::new(0x1, 11).unwrap()),
+            Ordering::Equal
+        );
+        assert_eq!(
+            FdtReserveEntry::new(0x1, 10)
+                .unwrap()
+                .cmp(&FdtReserveEntry::new(0x2, 10).unwrap()),
+            Ordering::Less
+        );
+        assert_eq!(
+            FdtReserveEntry::new(0x1, 10)
+                .unwrap()
+                .cmp(&FdtReserveEntry::new(0x0, 10).unwrap()),
+            Ordering::Greater
+        );
+    }
+
+    #[test]
+    fn test_overlapping_mem_reservations() {
+        // Check that regions that overlap return an error on new.
+        // Regions start exactly at the end location of the
+        let overlapping = [
+            FdtReserveEntry::new(0x3, 1).unwrap(), // this overlaps with
+            FdtReserveEntry::new(0x0, 1).unwrap(),
+            FdtReserveEntry::new(0x2, 2).unwrap(), // this one.
+        ];
+        let fdt = FdtWriter::new(&overlapping);
+        assert_eq!(fdt.unwrap_err(), Error::OverlappingMemoryReservations);
+    }
```

Comment:
It might also be good to have a test case for valid reservations that are directly contiguous without overlapping just to make sure no off-by-one errors sneak in later - e.g. {0, 1}, {1, 1}, {2, 1}.

Reply:
Sounds like a good idea, will add it.